File: aising2020/e/e.cc
#include <cstdio>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;
typedef long long ll;

template <typename T> void read(T &x) {
  x = 0;
  char c = getchar();
  T sig = 1;
  for (; !isdigit(c); c = getchar())
    if (c == '-')
      sig = -1;
  for (; isdigit(c); c = getchar())
    x = (x << 3) + (x << 1) + c - '0';
  x *= sig;
}

class Solution {
public:
  void solve() {
    int n;
    read(n);
    vector<vector<int>> red(n + 1), blue(n + 1);
    vector<int> L(n), R(n);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    ll ans = 0;
    for (int i = 0; i < n; ++i) {
      int k, l, r;
      read(k), read(l), read(r);
      if (l >= r)
        red[k].emplace_back(i);
      else
        blue[k].emplace_back(i);
      L[i] = l;
      R[i] = r;
      ans += min(l, r);
    }
    for (int i = 1; i <= n; ++i) {
      for (int j : red[i])
        pq.push({L[j] - R[j], j});
      while (pq.size() > i)
        pq.pop();
    }
    while (!pq.empty()) {
      ans += pq.top().first;
      pq.pop();
    }

    for (int i = n; i >= 1; --i) {
      for (int j : blue[i])
        pq.push({R[j] - L[j], j});
      while (pq.size() > n - i)
        pq.pop();
    }
    while (!pq.empty()) {
      ans += pq.top().first;
      pq.pop();
    }
    cout << ans << endl;
  }
};

int main() {
  ios::sync_with_stdio(false);
  cin.tie(0);
  int t;
  read(t);
  while (t--) {
    Solution solution = Solution();
    solution.solve();
  }
}

File: aising2020/e/e.kt
import java.util.*

fun readInt(): Int {
    return readLine()!!.toInt()
}

fun readInts(): List<Int> {
    return readLine()!!.split(" ").map(String::toInt)
}

class Solution {
    fun solve() {
        val n = readInt()
        val red = Array(n + 1) { mutableListOf<Int>() }
        val blue = Array(n + 1) { mutableListOf<Int>() }
        val l = IntArray(n)
        val r = IntArray(n)
        var ans = 0L
        for (i in 0 until n) {
            val (ki, li, ri) = readInts()
            if (li >= ri)
                red[ki].add(i)
            else
                blue[ki].add(i)
            l[i] = li
            r[i] = ri
            ans += li.coerceAtMost(ri)
        }

        val pq = PriorityQueue<Pair<Int, Int>>(compareBy<Pair<Int, Int>> { it.first }.thenBy { it.second })
        for (i in 1 until n + 1) {
            for (j in red[i])
                pq.add(Pair(l[j] - r[j], j))
            while (pq.size > i)
                pq.poll();
        }
        while (pq.isNotEmpty())
            ans += pq.poll().first

        for (i in n downTo 1) {
            for (j in blue[i])
                pq.add(Pair(r[j] - l[j], j))
            while (pq.size > n - i)
                pq.poll();
        }
        while (pq.isNotEmpty())
            ans += pq.poll().first

        println(ans)
    }
}

fun main() {
    val t = readInt()
    for (i in 1 until t + 1) {
        val solution = Solution()
        solution.solve()
    }
}
